frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx

```diff
@@ -1,62 +1,50 @@
-import { useCallback, useMemo, useState, useEffect } from "react";
+import { useCallback, useEffect, useMemo, useState } from "react";
+import { useLatest } from "react-use";
 import { t } from "ttag";
 import _ from "underscore";
 
 import EmptyState from "metabase/components/EmptyState";
 
-import { ActionsApi, PublicApi } from "metabase/services";
-
 import ActionForm from "metabase/actions/components/ActionForm";
-import { getDashboardType } from "metabase/dashboard/utils";
 
 import type {
-  WritebackParameter,
   OnSubmitActionForm,
-  Dashboard,
-  ActionDashboardCard,
   ParametersForActionExecution,
   WritebackAction,
+  WritebackParameter,
 } from "metabase-types/api";
 
 export interface ActionParametersInputFormProps {
   action: WritebackAction;
-  dashboard?: Dashboard;
-  dashcard?: ActionDashboardCard;
   mappedParameters?: WritebackParameter[];
-  dashcardParamValues?: ParametersForActionExecution;
+  initialValues?: ParametersForActionExecution;
+  fetchInitialValues?: () => Promise<ParametersForActionExecution>;
+  shouldPrefetch?: boolean;
   onSubmit: OnSubmitActionForm;
   onSubmitSuccess?: () => void;
   onCancel?: () => void;
 }
 
-const shouldPrefetchValues = (action: WritebackAction) =>
-  action.type === "implicit" && action.kind === "row/update";
-
 function ActionParametersInputForm({
   action,
   mappedParameters = [],
-  dashcardParamValues = {},
-  dashboard,
-  dashcard,
+  initialValues = {},
+  fetchInitialValues,
+  shouldPrefetch,
   onCancel,
   onSubmit,
-  onSubmitSuccess,
+  onSubmitSuccess = _.noop,
 }: ActionParametersInputFormProps) {
+  const onSubmitSuccessRef = useLatest(onSubmitSuccess);
```

Comment:
Any prop that is referenced in hooks (callback or effect) after the first `await` statement is prone to suffer from stale closure values. This can result in nasty bugs. 

This change also prevents bugs in future usages of this component.

There are few articles about it, here's one: https://css-tricks.com/dealing-with-stale-props-and-states-in-reacts-functional-components/

Reply:
I thought you encountered some bugs already. In this case, I think both `onSubmitSuccess` and `prefetchValues` (which depends on `fetchInitialValues`) should be safe to be used directly because:
1. `onSubmitSuccess` just call `onClose` and `onSuccess` which calls these functions that don't have any dependencies on them.
    - https://github.com/metabase/metabase/blob/6c568b904cd6a2bba738ee3fbd925e59ce976d0c/frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx#L338-L340
    - https://github.com/metabase/metabase/blob/6c568b904cd6a2bba738ee3fbd925e59ce976d0c/frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx#L165-L167
2. From the usage of `prefetchValues` it feels wrong that we could potentially use a newer thus different `prefetchValues` with older `onSubmit`. e.g. what is `onSubmit` is for one model but somehow `perfetchValues` is for another model since we've somehow navigate to the other model. 
    https://github.com/metabase/metabase/blob/6c568b904cd6a2bba738ee3fbd925e59ce976d0c/frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx#L87-L93

This kind of interaction seems like you should use the callback that's defined for each render cycle. If there's anything needed to be updated it will just get updated via the `useEffect` or props change anyway. The example would be you wait for the API to finish, then you close the current modal. It shouldn't be that while you're waiting for the API to finish you close the modal then open another model detail and the API finishes then it calls the latest `onSubmitSuccess` which closes the new modal you just opened.

I think the use case is different from that article you mentioned. TBH, I'm not 100% right now so I couldn't say for sure this is wrong but I have a hunch that this sounds wrong. Please feel free to discuss this.

tl;dr: I'd not use `useLatest` since I feel it could cause bugs not preventing it.

Reply:
> 1. `onSubmitSuccess` just call `onClose` and `onSuccess` which calls these functions that don't have any dependencies on them.

That was one of my points - the consumer of this component should not worry about this.
You're right - with the current state of the code it does not matter. But consumers of this component can change in the future and suddenly there might be a bug in the child component, that wasn't modified.

> 2. From the usage of `prefetchValues` it feels wrong that we could potentially use a newer thus different `prefetchValues` with older `onSubmit`. e.g. what is `onSubmit` is for one model but somehow `perfetchValues` is for another model since we've somehow navigate to the other model. 

It's the other way around - we need to use the latest `prefetchValues` because `prefetchValues` is possibly for another model by then.

> Please feel free to discuss this.

Perhaps it'd be most efficient to discuss this on a call.

Reply:
We had a call about it, here are notes and conclusions:
- it seems that both approaches are valid with the current state of code in this component and how it's used
- we think both approaches can be counter-exampled and shown to not work as expected (in some cases)
- we decided not to use `useLatest` for now because it hasn't been used in the code extensively (be consistent)
- we think it's good to talk this problem through during frontend guild meeting
  - we should have aforementioned counter-examples provided in an RFC
  - we don't know what the conclusion of that RFC would be

Reply:
Done: 9b779f0a28f393d4017c671d777621aa4ccb33ae